test/e2e_node/node_shutdown_linux_test.go

```diff
@@ -0,0 +1,217 @@
+	"k8s.io/kubernetes/pkg/apis/scheduling"
+	"k8s.io/kubernetes/pkg/features"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	kubelettypes "k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = framework.KubeDescribe("GracefulNodeShutdown [Serial] [NodeAlphaFeature:GracefulNodeShutdown]", func() {
+	f := framework.NewDefaultFramework("graceful-node-shutdown")
+	ginkgo.Context("when gracefully shutting down", func() {
+
+		const (
+			pollInterval                        = 1 * time.Second
+			podStatusUpdateTimeout              = 5 * time.Second
+			nodeStatusUpdateTimeout             = 10 * time.Second
+			nodeShutdownGracePeriod             = 20 * time.Second
+			nodeShutdownGracePeriodCriticalPods = 10 * time.Second
+		)
+
+		tempSetCurrentKubeletConfig(f, func(initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.FeatureGates = map[string]bool{
+				string(features.GracefulNodeShutdown): true,
+			}
+			initialConfig.ShutdownGracePeriod = metav1.Duration{Duration: nodeShutdownGracePeriod}
+			initialConfig.ShutdownGracePeriodCriticalPods = metav1.Duration{Duration: nodeShutdownGracePeriodCriticalPods}
+		})
+
+		ginkgo.It("should be able to gracefully shutdown pods with various grace periods", func() {
+			nodeName := getNodeName(f)
+			nodeSelector := fields.Set{
+				"spec.nodeName": nodeName,
+			}.AsSelector().String()
+
+			// Define test pods
+			pods := []*v1.Pod{
+				getGracePeriodOverrideTestPod("period-120", nodeName, 120, false),
+				getGracePeriodOverrideTestPod("period-5", nodeName, 5, false),
+				getGracePeriodOverrideTestPod("period-critical-120", nodeName, 120, true),
+				getGracePeriodOverrideTestPod("period-critical-5", nodeName, 5, true),
+			}
+
+			ginkgo.By("Creating batch pods")
+			f.PodClient().CreateBatch(pods)
+
+			list, err := f.PodClient().List(context.TODO(), metav1.ListOptions{
+				FieldSelector: nodeSelector,
+			})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				framework.ExpectEqual(
+					pod.Status.Phase == v1.PodRunning,
+					true,
+					"pod is not ready",
+				)
+			}
+
+			ginkgo.By("Emitting shutdown signal")
+			err = emitSignalPrepareForShutdown(true)
+			framework.ExpectNoError(err)
+
+			// Not critical pod should be shutdown
+			gomega.Eventually(func() error {
+				list, err = f.PodClient().List(context.TODO(), metav1.ListOptions{
+					FieldSelector: nodeSelector,
+				})
+				if err != nil {
+					return err
+				}
+				for _, pod := range list.Items {
+					if kubelettypes.IsCriticalPod(&pod) {
+						if pod.Status.Phase != v1.PodRunning {
+							return fmt.Errorf("critical pod should not be shutdown")
+						}
+					} else {
+						if pod.Status.Phase != v1.PodFailed || pod.Status.Reason != "Shutdown" {
+							return fmt.Errorf("pod should be shutdown")
+						}
+
+					}
+				}
+				return nil
+			}, podStatusUpdateTimeout, pollInterval).Should(gomega.BeNil())
```

Comment:
good point. Maybe one simple solution to start with is just to increase the `nodeShutdownGracePeriod` from `20s` and `nodeShutdownGracePeriodCriticalPods` from `10s` to something much higher, say `nodeShutdownGracePeriod`: `5min` and  `nodeShutdownGracePeriodCriticalPods`: `2min`

Having pods signal back would be ideal but I think it makes the test more complicated, maybe makes sense as a followup... I also brought this up in https://github.com/kubernetes/kubernetes/pull/98658#issuecomment-772172702

Reply:
I think the time to add is not very good. This will add a few minutes of checking time for each PR. 
Pods can use some logs to send out signals. I think it’s a good idea.

Reply:
if we want to proceed with this, can you add a comment suggesting that this may become flaky so it will be easier to investigate.

Reply:
I'm fine going as is and see if it will actually become a problem on e2e environment